h/services/user_delete.py

```diff
@@ -1,70 +1,315 @@
+            )
+
+    def delete_groups(self, user):
+        """
+        Delete groups created by `user` that have no annotations.
+
+        Delete groups that were created by `user` and that don't contain any
+        non-deleted annotations.
+
+        If delete_annotations() (above) is called first then all of `user`'s
+        own annotations will already have been deleted, so ultimately any
+        groups created by `user` that don't contain any annotations by *other*
+        users will get deleted.
+
+        Known issue: if this method does not delete a group because it contains
+        annotations by other users, and those annotations other users are later
+        deleted, then the group will no longer contain any annotations but will
+        never be deleted.
+
+        Known issue: this will also delete all of the groups memberships due to
+        a foreign key constraint with ondelete="cascade". If a group had a
+        really large number of members this could take too long and cause a
+        timeout.
+        """
+        # pylint:disable=not-callable,use-implicit-booleaness-not-comparison-to-zero
+        self.worker.delete(
+            Group,
+            select(Group.id)
+            .where(Group.creator_id == user.id)
+            .outerjoin(
+                Annotation,
+                and_(
+                    Annotation.groupid == Group.pubid,
+                    Annotation.deleted.is_(False),
+                ),
+            )
+            .group_by(Group.id)
+            .having(func.count(Annotation.id) == 0),
+        )
+
+    def delete_group_memberships(self, user):
+        """
+        Delete `user`'s group memberships.
+
+        Known issue: this can leave groups that were created by `user` in an
+        odd state - `user` will no longer be a member of the group but will
+        still be its creator. But this state can occur in other ways as well,
+        for example the web interface currently allows a group's creator to
+        leave the group.
+
+        If `delete_group_creators()` (below) is called after this method then
+        the situation will be only temporary: `user` will soon be removed as
+        the group's creator as well.
+        """
+        self.worker.delete(
+            GroupMembership,
+            select(GroupMembership.id)
+            .where(GroupMembership.user_id == user.id)
+            .join(Group, GroupMembership.group_id == Group.id),
+        )
+
+    def delete_group_creators(self, user):
+        """
+        Delete `user` as the creator of any groups they created.
+
+        Known issue: this will leave groups in an odd state - with no creator
+        (group.creator = None).
+        """
+        self.worker.update(
+            Group, select(Group.id).where(Group.creator == user), {"creator_id": None}
+        )
+
+    def delete_user(self, user):
+        """Delete `user`."""
+        self.worker.delete(User, select(User.id).where(User.id == user.id))
```

Comment:
A nice to have would be to have UserDeletion.completed_at and update it here.

Reply:
I think it would just be misleading unfortunately. At the time when the `user` row is deleted the user may actually still have annotations (and annotation slims etc) in the DB waiting to be purged by the `purge-deleted-annotations` task, and they may still have annotations in Elasticsearch waiting to be purged by the `sync-annotation` task.